### ISSUE TITLE
Renderer crash with amdgpu-pro OGLP driver on Linux, `nullptr` festival

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/Daemon#945]
When running proprietary amdgpu-pro OGLP driver version `23.20-1664987.22.04` (from amdgpu-pro `5.7.1`) on Ubuntu 23.10, I get a crash.

The `glContext` pointer is properly set to something that is not `nullptr` so the game believes everything is right and crash elsewhere.

For example, while `glContext` is not `nullptr` and is considered valid, doing things like `glGetString( GL_VERSION )` is returning a `nullptr`, if I add a workaround for this, I get some other `nullptr` error somewhere else in the code, like with `GL_RENDERER`.

What's weird is that the driver works, for example I can run Xonotic 0.8.6 without problem and the DarkPlaces engine prints:

```
GL_VENDOR: ATI Technologies Inc.
GL_RENDERER: AMD Radeon PRO W7600
GL_VERSION: 4.6.0 Compatibility Profile Context 23.20.230929
```

But more curiously, if I run an older prebuilt version of Unvanquished, like the `0.51.2`, the game runs and the engine prints:

```
GL_VENDOR: ATI Technologies Inc.
GL_RENDERER: AMD Radeon PRO W7600
GL_VERSION: 4.6.0 Compatibility Profile Context 23.20.230929
GL_SHADING_LANGUAGE_VERSION: 4.60
```

The proprietary OGLP driver is not recommended by AMD (they recommend the stock free open source Mesa driver instead), unless you know why you need it. Anyway, it is distributed withing AMDGPU-PRO package, and then people may use it, either intentionally, either by mistake.

Since other games run, and especially older versions of the game, it's possible we may be doing something wrong.

**Comment by illwieckz** [comment]
Our current release (`0.54.0`) works… but

1. It was not built on the same system.
2. We have updated SDL2.

**Comment by illwieckz** [comment]
A release build from `master` runs properly, the release build has a statically self-built SDL2, so maybe it's a bug in stock SDL2 from Ubuntu 23.10?

**Comment by illwieckz** [comment]
Building on Ubuntu 23.10 but with our own SDL2 external_deps reproduces the bug, so the stock Ubuntu SDL2 is probably not the culprit.

**Comment by illwieckz** [comment]
For reference, here is the backtrace:

```gdb
Thread 1 (Thread 0x7ffff5500a80 (LWP 859175) "daemon"):
#0  __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:76
#1  0x00007ffff7092674 in _IO_str_init_static_internal (sf=sf@entry=0x7fffffffcd10, ptr=ptr@entry=0x0, size=size@entry=0, pstart=pstart@entry=0x0) at ./libio/strops.c:41
#2  0x00007ffff705c222 in _IO_strfile_read (string=0x0, sf=0x7fffffffcd10) at ../libio/strfile.h:90
#3  __GI___isoc23_sscanf (s=0x0, format=format@entry=0x5555557a2c5a "%d.%d") at ./stdio-common/isoc23_sscanf.c:28
#4  0x000055555570fafc in GLimp_RegisterConfiguration (highestConfiguration=..., requestedConfiguration=...) at Unvanquished/daemon/src/engine/sys/sdl_glimp.cpp:1240
#5  GLimp_SetMode (mode=<optimized out>, fullscreen=fullscreen@entry=false, bordered=bordered@entry=false) at Unvanquished/daemon/src/engine/sys/sdl_glimp.cpp:1422
#6  0x0000555555710759 in GLimp_StartDriverAndSetMode (mode=<optimized out>, fullscreen=false, bordered=false) at Unvanquished/daemon/src/engine/sys/sdl_glimp.cpp:1521
#7  0x00005555557109df in GLimp_Init () at Unvanquished/daemon/src/engine/sys/sdl_glimp.cpp:1951
#8  0x00005555556c1e8e in InitOpenGL () at Unvanquished/daemon/src/engine/renderer/tr_init.cpp:307
#9  0x00005555556c3b60 in R_Init () at Unvanquished/daemon/src/engine/renderer/tr_init.cpp:1414
#10 0x00005555556d2f23 in RE_BeginRegistration (glconfigOut=0x555555d7cdb4 <cls+2802484>, glconfig2Out=0x555555d7d9e0 <cls+2805600>) at Unvanquished/daemon/src/engine/renderer/tr_model.cpp:259
#11 0x0000555555616b12 in CL_InitRenderer () at Unvanquished/daemon/src/engine/client/cl_main.cpp:2134
#12 0x0000555555616dd5 in CL_StartHunkUsers () at Unvanquished/daemon/src/engine/client/cl_main.cpp:2208
#13 CL_StartHunkUsers () at Unvanquished/daemon/src/engine/client/cl_main.cpp:2194
#14 0x000055555578686f in Sys::Init (argc=<optimized out>, argv=0x7fffffffdc38) at Unvanquished/daemon/src/engine/framework/System.cpp:721
#15 0x00005555555a24df in main (argc=<optimized out>, argv=<optimized out>) at Unvanquished/daemon/src/engine/framework/System.cpp:741
```

**Comment by illwieckz** [comment]
This may not be a bug in dæmon, maybe we face some library mismatch between the driver and daemon, but it's still useful to document this.

It looks like @cu-kai is reproducing a very similar bug [when running a custom-built Mesa on Arch and using some Wayland/Prime options](https://gitlab.freedesktop.org/mesa/mesa/-/issues/9282#note_2187658):

```gdb


Thread 1 "daemon" received signal SIGSEGV, Segmentation fault.
0x00007ffff7178edd in ?? () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff7178edd in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff70a5b85 in ?? () from /usr/lib/libc.so.6
#2  0x00007ffff7073d34 in __isoc23_sscanf () from /usr/lib/libc.so.6
#3  0x00005555556b6992 in GLimp_RegisterConfiguration (highestConfiguration=..., requestedConfiguration=...) at daemon/src/engine/sys/sdl_glimp.cpp:1240
#4  0x00005555556b6fcd in GLimp_SetMode (mode=<optimized out>, fullscreen=fullscreen@entry=false, bordered=bordered@entry=true) at daemon/src/engine/sys/sdl_glimp.cpp:1422
#5  0x00005555556b7253 in GLimp_StartDriverAndSetMode (mode=<optimized out>, fullscreen=false, bordered=bordered@entry=true) at daemon/src/engine/sys/sdl_glimp.cpp:1521
#6  0x00005555556b7e12 in GLimp_Init () at daemon/src/engine/sys/sdl_glimp.cpp:1951
#7  0x0000555555676039 in InitOpenGL () at daemon/src/engine/renderer/tr_init.cpp:307
#8  0x000055555567894c in R_Init () at daemon/src/engine/renderer/tr_init.cpp:1414
#9  0x0000555555686bd5 in RE_BeginRegistration (glconfigOut=0x555555d28714 <cls+2802484>, glconfig2Out=0x555555d29340 <cls+2805600>) at daemon/src/engine/renderer/tr_model.cpp:259
#10 0x00005555555e424c in CL_InitRenderer () at daemon/src/engine/client/cl_main.cpp:2135
#11 0x00005555555e44b8 in CL_StartHunkUsers () at daemon/src/engine/client/cl_main.cpp:2209
#12 0x000055555557c78f in Application::ClientApplication::Initialize (this=<optimized out>) at daemon/src/engine/client/ClientApplication.cpp:92
#13 0x00005555556ee55f in Application::Initialize () at daemon/src/engine/framework/Application.cpp:71
#14 0x000055555570a3d9 in Sys::Init (argc=16, argv=0x7fffffffe2d8) at daemon/src/engine/framework/System.cpp:720
#15 0x000055555570a53b in main (argc=<optimized out>, argv=<optimized out>) at daemon/src/engine/framework/System.cpp:740
```

**Comment by illwieckz** [comment]
I summon @slipher as I would need some master wizard opinion on such dark magic bug.

I actually suspect some library mismatches between the driver and the game: maybe the game or SDL2 is built against a version of a library, but the driver already loaded an incompatible variant of that library, or something like that?

**Comment by slipher** [comment]
Yeah updating GLEW would be a good thing to try if there is a new version. On Mac there was a problem that GLEW started to hand us some null function pointers, which was fixed by updating GLEW.

**Comment by illwieckz** [comment]
I get this error when I build the engine with Ubuntu 23.10 system's GLEW 2.2.0

The Debian buster static build with non-system GLEW 2.2.0 we do with docker doesn't suffer from that bug on my end.

**Comment by illwieckz** [comment]
Building against our own GLEW with those external deps:

```
./build.sh linux-amd64-default glew naclsdk install
```

Brings me the same bug…

**Comment by illwieckz** [comment]
I don't know if that's related, but building GLEW 2.2.0 on Ubuntu 23.10 with GCC 13.2.0 brings me those warnings:

```
src/glew.c: In function ‘_glewStrSame1’:
src/glew.c:256:29: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*a + (sizetype)i’ must not be NULL [-Waddress]
  256 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                             ^~
src/glew.c:256:44: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘b + (sizetype)i’ must not be NULL [-Waddress]
  256 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                                            ^~
src/glew.c: In function ‘_glewStrSame2’:
src/glew.c:272:29: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*a + (sizetype)i’ must not be NULL [-Waddress]
  272 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                             ^~
src/glew.c:272:44: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘b + (sizetype)i’ must not be NULL [-Waddress]
  272 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                                            ^~
src/glew.c: In function ‘_glewStrSame3’:
src/glew.c:288:29: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*a + (sizetype)i’ must not be NULL [-Waddress]
  288 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                             ^~
src/glew.c:288:44: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘b + (sizetype)i’ must not be NULL [-Waddress]
  288 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                                            ^~
src/glew.c: In function ‘_glewStrSame1’:
src/glew.c:256:29: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*a + (sizetype)i’ must not be NULL [-Waddress]
  256 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                             ^~
src/glew.c:256:44: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘b + (sizetype)i’ must not be NULL [-Waddress]
  256 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                                            ^~
src/glew.c: In function ‘_glewStrSame2’:
src/glew.c:272:29: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*a + (sizetype)i’ must not be NULL [-Waddress]
  272 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                             ^~
src/glew.c:272:44: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘b + (sizetype)i’ must not be NULL [-Waddress]
  272 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                                            ^~
src/glew.c: In function ‘_glewStrSame3’:
src/glew.c:288:29: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*a + (sizetype)i’ must not be NULL [-Waddress]
  288 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |                             ^~
src/glew.c:288:44: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘b + (sizetype)i’ must not be NULL [-Waddress]
  288 |     while (i < nb && (*a)+i != NULL && b+i != NULL && (*a)[i] == b[i]) i++;
      |  
```

**Comment by slipher** [comment]
When building your own deps are you sure that you are actually using them? You probably have to manually set the cmake variables (in the release script they are only used automatically because we set USE_STATIC_LIBS or whatever.) The `slipher/printdepversions` branch might be interesting to make sure.

**Comment by illwieckz** [comment]
Both `CMake` and `ldd` tells me I use my custom GLEW.

**Comment by slipher** [comment]
Nice find! I don't remember why I had to put `-DOpenGL_GL_PREFERENCE=LEGACY` there (compatibility with old systems maybe) but apparently it's a good thing

**Comment by illwieckz** [comment]
@cu-kai confirmed on IRC that building with `-DOpenGL_GL_PREFERENCE=LEGACY` also fixes the segfaults he gets when running the engine on Wayland with Prime:

```
2023-11-30 19:59:06 +0100 <illwieckz> cu-kai, what if you build daemon with -DOpenGL_GL_PREFERENCE=LEGACY
2023-11-30 20:15:24 +0100 <cu-kai> how do i compile with this option?
2023-11-30 20:15:51 +0100 <illwieckz> cd build ; cmake -DOpenGL_GL_PREFERENCE=LEGACY .. ; make -j$(nproc)
2023-11-30 20:16:03 +0100 <cu-kai> i did exactly this
2023-11-30 20:16:26 +0100 <cu-kai> fixes for me too
```

**Comment by illwieckz** [comment]
I renamed this thread as “_Renderer segfault with GLVND OpenGL ABI on Linux, `nullptr` festival_”, so this thread is not about the specific OGLP use case that triggers that GLVND issue, but about this family of GLVND issues in general.

**Comment by illwieckz** [comment]
It looks like the reason why we use `-DOpenGL_GL_PREFERENCE=LEGACY` in release builds is that despite GLVND is something introduced by Nvidia, they never re-released their legacy drivers so it seems that only engines build with LEGACY can run on older Nvidia drivers and then older Nvidia hardware (if nouveau is not performant enough).

Unless I miss something, it looks like building with GLVND would immediately make the engine incompatible with a whole range of hardware and drivers that are totally capable to run the game:

- https://github.com/DaemonEngine/Daemon/issues/474

Here, @necessarily-equal said that “Not having a GLX context is not a bug: wayland uses EGL for example”, I wonder if that's related:

- https://github.com/DaemonEngine/Daemon/pull/483

**Comment by illwieckz** [comment]
So, here are quotes from `#dri-devel` @ OFTC (basically the historical Mesa developer channel).

TL;DR:

- Building with a binary with `LEGACY` ABI is compatible with GLVND-aware `libGL.so` and legacy `libGL.so`.
- Building a binary with `GLVND` ABI is incompatible with `libGL.so`, it uses backward-incompatible new `libOpenGL.so` instead.
- A software developer shouldn't have to use `GLVND` ABI (and explicitly target backward-incompatible `libOpenGL.so`) unless he wants to and knows why he explictly needs such backward-incompatible ABI.
- CMake is likely breaking people's apps in unknowingly developer's back and against their will by setting `GLVND` the default ABI.

The quotes:

```
2023-12-01 00:49:59 +0100 <illwieckz> Do someone knows a bit about glvnd? What
                          problems is it trying to solve? Unless we do something
                          wrong with the Dæmon game engine, accumulated
                          experience over two years is that building an OpenGL
                          application with GLVND ABI instead of LEGACY one
                          1. breaks compatibility with legacy Nvidia drivers,
                          2. breaks compatibility with AMD OGLP driver,
                          3. breaks compatibility with Wayland with Prime…
2023-12-01 00:53:00 +0100 <illwieckz> Is glvnd expected to be backward
                          compatible and all of these problems are bugs, or is
                          it considered OK that maybe only current Nvidia
                          drivers on Xorg is expected to work and we are just
                          lucky if current Mesa drivers are compatible with it
                          on Xorg too ?


2023-12-01 00:54:18 +0100 <Kayden> so there are two aspects to glvnd
2023-12-01 00:55:08 +0100 <Kayden> first is just the new vendor neutral dispatch
                          library.  glvnd provides libGL.so, and it loads the
                          vendor drivers (libGLX_mesa.so, libEGL_mesa.so,
                          libGLX_nvidia.so, etc)
2023-12-01 00:55:36 +0100 <Kayden> that should be fully backwards compatible,
                          and is meant as a way to let you parallel install mesa
                          drivers, nvidia proprietary drivers, fglrx (not that
                          that's a thing anymore) & whatever else
2023-12-01 00:56:06 +0100 <Kayden> as long as you're linking against libGL.so it
                          should not matter whether it's glvnd or mesa or nvidia
                          directly

2023-12-01 00:56:18 +0100 <Kayden> but it sounds like you may be running into
                          the second aspect
2023-12-01 00:56:30 +0100 <Kayden> which is...glvnd decided to define a new ABI
                          for OpenGL, libOpenGL.so
2023-12-01 00:56:45 +0100 <Kayden> which is not backwards compatible
2023-12-01 00:57:00 +0100 <Kayden> IIRC it only has OpenGL core profile support.
                          It also does not contain GLX
2023-12-01 00:57:12 +0100 <Kayden> libGL has both OpenGL core, OpenGL
                          compatibility (legacy), and GLX
2023-12-01 00:57:28 +0100 <Kayden> which is awkward if you're trying to do, say,
                          EGL only, on a non-X11 windowing environment
2023-12-01 00:57:54 +0100 <illwieckz> so, does it mean that I can build an app
                          with -DOpenGL_GL_PREFERENCE=LEGACY
2023-12-01 00:58:05 +0100 <illwieckz> but use a glvnd libGL.so ?
2023-12-01 00:58:16 +0100 <Kayden> I've not heard of that flag, but yes, it
                          should be possible
2023-12-01 00:58:33 +0100 <Kayden> you shouldn't have to target libOpenGL.so as
                          an app / engine author unless you -want- to

2023-12-01 00:58:37 +0100 <illwieckz> that's the cmake flag for selecting either
                          LEGACY ABI or GLVND ABI
2023-12-01 00:58:48 +0100 <illwieckz> ok, good to know
2023-12-01 00:59:01 +0100 <illwieckz> because CMake defaults to
                          -DOpenGL_GL_PREFERENCE=GLVND
2023-12-01 00:59:31 +0100 <illwieckz> even if the developer has no idea this
                          will break backward compatibility if I understand it
                          right
2023-12-01 01:00:13 +0100 <Kayden> yeah, I think that would break compatibility
                          with super legacy drivers that never transitioned

2023-12-01 01:03:17 +0100 <illwieckz> so, I assume the good thing to know for a
                          developer is: don't use GLVND ABI unless you know why
                          you need it
2023-12-01 01:04:20 +0100 <illwieckz> Maybe we may report to CMake that setting
                          the GLVND ABI the default is probably a mistake.
```

**Comment by illwieckz** [comment]
With the acquired knowledge, I now believe we need to use `LEGACY` by default and that CMake using `GLVND` by default is a mistake from CMake.

So I added a commit to use `LEGACY` OpenGL ABI on Linux by default unless `OpenGL_GL_PREFERENCE` is explictely set (and setting it to empty string is expected to use CMake default instead) to that PR:

- https://github.com/DaemonEngine/Daemon/pull/976
